File: leetcode/golang/20_valid_parentheses.go
func isValid(s string) bool {
    stack := []byte{}
    for i := 0; i < len(s); i++ {
        if (len(stack) == 0 && (s[i] == ')' || s[i] == ']' || s[i] == '}')) {
            return false
        }
        if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
            stack = append(stack, s[i])
        } else if (stack[len(stack) - 1] == '(' && s[i] == ')' || stack[len(stack) - 1] == '[' && s[i] == ']' || stack[len(stack) - 1] == '{' && s[i] == '}') {
            stack = stack[: len(stack) - 1]
        } else {
            return false;
        }   
    }
    return len(stack) == 0
}

File: leetcode/c++/20_valid_parentheses.cpp
// https://leetcode.cn/problems/valid-parentheses/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
       stack<char>S;
       for (char ch : s) {
            if (S.empty() && (ch == ')' || ch == ']' || ch == '}'))
                return false;
            if (ch == '(' || ch == '[' || ch == '{')
                S.push(ch);
            else if (S.top() == '(' && ch == ')' || S.top() == '[' && ch == ']' || S.top() == '{' && ch == '}')
                S.pop();
            else
                return false;
       }
       return S.empty();
    }
};

// 时间复杂度和空间复杂度都是O(n)